Update api.md (#459)
Codecademy/docs: content/general/concepts/api/api.md

---
Title: 'API'

> Description: 'API or Application Programing Interface is a term used to describe specifications that allow applications to communicate with one another. APIs enable exchange of information and can be a major source of value to organizations. APIs can be divided into three groups: Public APIs, Private APIs, and Partner APIs.'

Reviewer: ```suggestion
Description: 'An API, or Application Programming Interface, is a term used to describe specifications that allow applications to communicate with one another. APIs enable exchange of information and can be a major source of value to organizations. They can be divided into three groups: Public APIs, Private APIs, and Partner APIs.'
```

- Private APIs, also known as Internal APIs, are used primarily within a company to share resources and facilitate the business. (e.g. Company Warehouse API for managing inventory with code).

- Partner APIs require rights or specific licenses for use. These APIs are popular in software-as-a-service platforms (e.g. [AWS API](https://docs.aws.amazon.com/general/latest/gr/aws-apis.html)).

## Benefits and Risks of Using APIs

The main benefits of using APIs are:

- Cost savings through the elimination of the need to build and maintain local physical infrastructure and services.

> - Opportunities to leverage the expertise of other vendors instead of having to reinvent solutions for common problems. (authentication, payment processing, maps, etc.)

Reviewer 2: Can we make "(authentication, payment processing, maps, etc.)" its own sentence by explicitly stating these are some examples, or if we want to keep it in parenthesis we should move it inside the sentence?

Author: Should be inside the sentence, I guess...

- Private APIs, also known as Internal APIs, are used primarily within a company to share resources and facilitate the business. (e.g. Company Warehouse API for managing inventory with code).

- Partner APIs require rights or specific licenses for use. These APIs are popular in software-as-a-service platforms (e.g. [AWS API](https://docs.aws.amazon.com/general/latest/gr/aws-apis.html)).

## Benefits and Risks of Using APIs

The main benefits of using APIs are:

> - Cost saving through the elimination of the need to build and maintain local physical infrastructure and services.

Reviewer: ```suggestion
- Savings of cost through the elimination of the need to build and maintain local physical infrastructure and services.
```

> - Opportunity to leverage the expertise of different vendors instead of having to reinvent from scratch things like authentication, payment processing, maps, etc.

Reviewer: ```suggestion
- Opportunity to leverage the expertise of other vendors instead of having to reinvent solutions for common problems. (authentication, payment processing, maps, etc.)
```

Author: OK!

Reviewer: If this is ok with you, go ahead and click "Commit suggestion", then click "Commit changes".

> - Possibility to build new businesses based on the exchange of data between users and various online services offering API access to independent developers.

Author: This is fine

The main risks associated with using APIs are:

> - Having no control over the vendor's infrastrucure or how the vendor allows developers to access it through the API.
> 
> - Unpredictability of the vendor's internal decision-making may have a very adverse effect on businesses relying on the API.
> 
> - Having little or no control over exactly what data the API collects and how it ends up being used by the API's provider.

Reviewer: ```suggestion
- Updates to an API may cause breaking changes for a product that uses it.
```

I feel like the first three list items all relate to how using an outside vendor's API leaves one at the mercy of any breaking changes that come with updates.

Author: I am not sure about this one. Updates are one thing, but the problem is that many more things may happen, including the vendor deciding to change their business model entirely or to misuse the data that is collected through the API.

Reviewer: Gotcha. I guess when I was reading through these three points, the common things that stuck out to me was these all have to do with what the vendor does with the API ("Having little or no control...", "Unpredictability...", etc.). 

Instead, what do we think of this:

```
- Little to no control over the impact of vendor-only changes including business infrastructure, version updates, and data collection. 
```

Author: Yes, this looks good!

Reviewer: Cool! Let's commit it.

> - APIs are gateways and are often subject to hacker attacks. As soon as an API is breached, other systems become vulnerable. It is commonly predicted that API abuses will become the most frequent vector of attack in the coming years.

Reviewer: ```suggestion
- Nonsecure APIs can become gateways to hacker attacks. As soon as it is breached, other dependent systems become vulnerable. It is commonly predicted that API abuses will become the most frequent vector of attack in the coming years.
```

Author: This is OK, I guess... Actually, I wanted to emphasize the fact that APIs are gateways (or 'doors') by their very nature, and it should be expected that hackers will try to break in through them. As I understand, this happens all the time. But anyway, what you wrote is also fine.

> - API providers run the risk of their legitimate business partners abusing the API to suit their own needs.

Reviewer: ```suggestion
- APIs can be abused by the clients themselves. One example is excessive requests to the API, which is what rate limits are meant to prevent.
```

Author: I was thinking of Cambridge Analytica and Facebook :-)

Reviewer: Lol let's not name names here 😉
